### PR TITLE
Certificate Revocation Validation with CRL and OCSP

### PR DESCRIPTION
**Pull request by IndunilRathnayake**
## Purpose
Including Certificate Revocation Validation with CRL and OCSP for X509 Authenticator

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java [hunk 1]
```diff
@@ -54,4 +54,7 @@ private X509CertificateConstants() {
     public static final String USERNAME_NOT_FOUND_FOR_X509_CERTIFICATE_ATTRIBUTE = "18003";
     public static final String X509_CERTIFICATE_USERNAME = "X509CertificateUsername";
     public static final String USER_NOT_FOUND = "17001";
+    public static final String X509_CERTIFICATE_NOT_VALID_ERROR_CODE = "18015";
+    public static final String X509_CERTIFICATE_NOT_VALIDATED_ERROR_CODE = "17003";
+
 }
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java:60]: Add a new line

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 2]
```diff
@@ -77,9 +85,9 @@ public static X509Certificate getCertificate(String username) throws Authenticat
                 throw new AuthenticationFailedException("Cannot find the user realm for the given tenant domain : " +
                         CarbonContext.getThreadLocalCarbonContext().getTenantDomain());
             }
-        } catch (javax.security.cert.CertificateException e) {
+        } catch (CertificateException e) {
             throw new AuthenticationFailedException("Error while decoding the certificate ", e);
-        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+        } catch (UserStoreException e) {
             throw new AuthenticationFailedException("Error while user manager for tenant id ", e);
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Add meaningful error message.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 3]
```diff
@@ -114,8 +124,6 @@ public static boolean addCertificate(String username, byte[] certificateBytes)
         } catch (CertificateException e) {
             throw new AuthenticationFailedException("Error while retrieving certificate of user: " + username, e);
         } catch (UserStoreException e) {
-            throw new AuthenticationFailedException("Error while setting certificate of user: " + username, e);
-        } catch (org.wso2.carbon.user.api.UserStoreException e) {
             throw new AuthenticationFailedException("Error while user manager for tenant id", e);
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: "Error while user manager for tenant id" is not meaningful . Hence change the error message as "Error while retrieving the user store manager for tenant id  " or as something meaningful.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 4]
```diff
@@ -212,9 +241,91 @@ public static UserRealm getUserRealm(String username) throws AuthenticationFaile
...
+                (userName);
+        if (isUserExist) {
+            return true;
+        } else {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.USER_NOT_FOUND);
+            throw new AuthenticationFailedException(" Unable to find X509 Certificate's user in user store. ");
+        }
+    }
 }
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Add new line

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 5]
```diff
@@ -212,9 +241,91 @@ public static UserRealm getUserRealm(String username) throws AuthenticationFaile
...
+                log.debug("Provided X509 client certificate has been revoked. Removing the x509Certificate claim " +
+                        "of the user.");
+            }
+            deleteCertificate(userName);
+        }
+    }
+
+    private static void deleteCertificate(String username)
+            throws AuthenticationFailedException {
+
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Formatting issue.. can be in same line

File: pom.xml [hunk 6]
```diff
@@ -365,8 +380,12 @@
     </distributionManagement>
     <properties>
         <carbon.identity.version>5.0.7</carbon.identity.version>
+        <carbon.identity.framework.version>5.7.5</carbon.identity.framework.version>
```
Comment (Kanapriya) [inline pom.xml]: Do we need both version 5.0.7 (old) and 5.7.5 (new framework) in the pom?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 7]
```diff
@@ -125,27 +133,55 @@ public static boolean addCertificate(String username, byte[] certificateBytes)
...
-        boolean validateResult;
         try {
-            x509Certificate = X509Certificate.getInstance(certificateBytes);
-            validateResult = x509Certificate.equals(getCertificate(userName));
-        } catch (javax.security.cert.CertificateException e) {
+            CertificateFactory cf = CertificateFactory.getInstance("X509");
+            x509Certificate = (X509Certificate) cf.generateCertificate
+                    (new ByteArrayInputStream(certificateBytes));
+
+            if (isCertificateRevoked(x509Certificate)) {
```
Comment (IndunilRathnayake) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: first check the user claim and then the revocation status

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 8]
```diff
@@ -212,9 +248,95 @@ public static UserRealm getUserRealm(String username) throws AuthenticationFaile
...
+        return revocationValidationManager.verifyRevocationStatus(x509Certificate);
+    }
+
+    private static void deleteUserCertificate(String userName, X509Certificate x509Certificate)
+            throws AuthenticationFailedException {
+
+        if (isCertificateExist(userName) && isUserCertificateValid(userName, x509Certificate)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Provided X509 client certificate has been revoked. Removing the x509Certificate claim " +
+                        "of the user.");
```
Comment (IndunilRathnayake) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: more info to logs. - username

**Review by Kanapriya — APPROVED** [review]

**Review by malithie — APPROVED** [review]

**Pull request merged**